bootstrap: improvements for compiler builds

onur-ozkan opened the pull request:
Reverted https://github.com/rust-lang/rust/pull/108288 and applied a proper fix with the following commit.

r? @Mark-Simulacrum

onur-ozkan committed:
revert temporary patch #108288

Signed-off-by: onur-ozkan <work@onurozkan.dev>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7557549416/job/20577041276) [(plain)](https://github.com/rust-lang/rust/commit/3e7f9ddb63a282cd48e8863be45b28c7f737aec6/checks/20577041276/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=onur-ozkan
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_9905795c-f57a-4eaf-a33f-977b2beaaf51
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=cb226a9ba3e16a1a2cdbdb1b9e2e4351848c0cf9
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_9905795c-f57a-4eaf-a33f-977b2beaaf51
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_9905795c-f57a-4eaf-a33f-977b2beaaf51
GITHUB_TRIGGERING_ACTOR=onur-ozkan
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/120058/merge
GITHUB_WORKFLOW_SHA=cb226a9ba3e16a1a2cdbdb1b9e2e4351848c0cf9
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---
Built container sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Looks like docker image is the same as before, not uploading
https://ci-caches.rust-lang.org/docker/7ebc15c01a233894034d277c8cce4e949f4e7791f66b4727c8fb6e058a0b8171d6152e1441d677cef0653843ceeee469c097b8699b2bb74249e674f6aa1a8813
sha256:9c3c93a371e5aed5c18185b24f130d95d5140dbd72a9b325e7b6b49e521a4faa
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Wed Jan 17 14:57:22 UTC 2024
  network time: Wed, 17 Jan 2024 14:57:22 GMT
  network time: Wed, 17 Jan 2024 14:57:22 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[endgroup]
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished release [optimized] target(s) in 1.35s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --no-default-features --mini-tests --std-tests`
Using system GCC
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   Compiling bootstrap v0.0.0 (/checkout/src/bootstrap)
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:227:17
    |
227 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 0 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:228:17
    |
    |
228 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 1 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:229:17
    |
    |
229 |                 compile::Assemble { target_compiler: Compiler { host: b, stage: 1 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:486:17
    |
    |
486 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 0 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:487:17
    |
    |
487 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 1 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:488:17
    |
    |
488 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 2 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:489:17
    |
    |
489 |                 compile::Assemble { target_compiler: Compiler { host: b, stage: 2 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:543:17
    |
    |
543 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 0 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:544:17
    |
    |
544 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 1 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
error[E0063]: missing field `host` in initializer of `Assemble`
##[error]   --> src/core/../tests/builder.rs:545:17
    |
    |
545 |                 compile::Assemble { target_compiler: Compiler { host: a, stage: 2 } },
    |                 ^^^^^^^^^^^^^^^^^ missing `host`
For more information about this error, try `rustc --explain E0063`.
error: could not compile `bootstrap` (lib test) due to 10 previous errors
Build completed unsuccessfully in 0:33:46
  local time: Wed Jan 17 15:31:43 UTC 2024
```

</details>

onur-ozkan committed:
distribute actual stage of the compiled compiler

By "actual" we refer to the uplifting logic where we may not compile the requested stage;
instead, we uplift it from the previous stages. Which can lead to bootstrap failures in
specific situations where we request stage X from other steps. However we may end up
uplifting it from stage Y, causing the other stage to fail when attempting to link with
stage X which was never actually built.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

onur-ozkan commented:
@rustbot ready

Mark-Simulacrum commented:
@bors r+ 

Doesn't seem obviously wrong to me.

bors commented:
:pushpin: Commit 8a461aa8cb4dc0758d605eb9240f2bd563267a9b has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 8a461aa8cb4dc0758d605eb9240f2bd563267a9b -->
<!-- homu: {"type":"Approved","sha":"8a461aa8cb4dc0758d605eb9240f2bd563267a9b","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
